pkg/gc/finalizer.go

```diff
@@ -150,32 +152,48 @@ func (f *subscriptionFinalizer) checkGeneration(ctx context.Context, typ core_mo
 		}
 
 		old, ok := f.insights[tenantID(tenantId)][typ][key]
-		if !ok || old.id != insight.GetLastSubscription().GetId() || old.generation != insight.GetLastSubscription().GetGeneration() {
+		if !ok || old.id != insight.GetLastSubscription().GetId() {
 			// something changed since the last check, either subscriptionId or generation were updated
 			// don't finalize the subscription, update map with fresh data
 			f.insights[tenantID(tenantId)][typ][key] = &descriptor{
-				id:         insight.GetLastSubscription().GetId(),
-				generation: insight.GetLastSubscription().GetGeneration(),
+				id: insight.GetLastSubscription().GetId(),
 			}
 			continue
 		}
 
-		log.V(1).Info("mark subscription as disconnected")
-		insight.GetLastSubscription().SetDisconnectTime(now)
-
-		upsertInsight, _ := registry.Global().NewObject(typ)
-		err := manager.Upsert(ctx, f.rm, key, upsertInsight, func(r core_model.Resource) error {
-			return upsertInsight.GetSpec().(generic.Insight).UpdateSubscription(insight.GetLastSubscription())
-		})
-		if err != nil {
-			log.Error(err, "unable to finalize subscription")
+		if exist, err := f.instanceExists(ctx, insight.GetLastSubscription().GetInstanceId()); err != nil {
 			return err
+		} else if !exist {
+			log.V(1).Info("mark subscription as disconnected")
+			insight.GetLastSubscription().SetDisconnectTime(now)
+
+			upsertInsight, _ := registry.Global().NewObject(typ)
+			err = manager.Upsert(ctx, f.rm, key, upsertInsight, func(r core_model.Resource) error {
+				return upsertInsight.GetSpec().(generic.Insight).UpdateSubscription(insight.GetLastSubscription())
+			})
+			if err != nil {
+				log.Error(err, "unable to finalize subscription")
+				return err
+			}
+			delete(f.insights[tenantID(tenantId)][typ], key)
 		}
-		delete(f.insights[tenantID(tenantId)][typ], key)
 	}
 	return nil
 }
 
+func (f *subscriptionFinalizer) instanceExists(ctx context.Context, id string) (bool, error) {
+	instances, err := f.catalog.Instances(ctx)
```

Comment:
Feels like this should be a helper in the catalog (or in the catalog package at least)

Reply:
```
func InstanceOfID(ctx context.Context, catalog Catalog, id string) (Instance, error) {
	instances, err := catalog.Instances(ctx)
	if err != nil {
		return Instance{}, err
	}
	for _, instance := range instances {
		if instance.Id == id {
			return instance, nil
		}
	}
	return Instance{}, ErrInstanceNotFound
}
```
there is this ^